### ISSUE TITLE
CLI 9.16.3 + Snowflake : ERROR: Incomplete statement at line

### ISSUE THREAD

**Issue by frichtarik** [flyway/flyway#3621]
###### Which version and edition of Flyway are you using?
9.16.3

###### If this is not the latest version, can you reproduce the issue with the latest one as well? (Many bugs are fixed in newer releases and upgrading will often resolve the issue)

yes, it's not happening with 9.11.0

###### Which client are you using? (Command-line, Java API, Maven plugin, Gradle plugin)

CLI 

###### Which database are you using? (Type & version)

Snowflake 7.10

###### Which operating system are you using?

Platform: Linux-4.4.0-204-generic-x86_64-with-glibc2.23

###### What did you do? (Please include the content causing the issue, any relevant configuration settings, the SQL statement(s) that failed (if any), and the command you ran)

same test procedure as for last few years (apply migrations to clean fresh DB), i just updated flyway

command i use (sensitive info replaced):

`/var/lib/jenkins/workspace/<JOB_NAME>/flyway/flyway migrate -schemas=CORE -url=jdbc:snowflake://<account>.snowflakecomputing.com:443/?ssl=on\&warehouse=<WH_NAME>\&db=frichtarik_<context>_20230210_225520\&schema=core\&role=<ROLE_NAME> -locations=filesystem:/migrations/core -placeholderPrefix=@{ -placeholderSuffix=}`

###### What did you expect to see?

success  (creation of schema and application of migrations)

###### What did you see instead?

INFO:root - QUOTED_IDENTIFIERS_IGNORE_CASE option is false
INFO:root - Flyway Community Edition 9.14.1 by Redgate
INFO:root - See what's new here: https://flywaydb.org/documentation/learnmore/releaseNotes#9.14.1
INFO:root - 
INFO:root - Database: jdbc:snowflake://<account>.snowflakecomputing.com:443/ (Snowflake 7.4)
INFO:root - QUOTED_IDENTIFIERS_IGNORE_CASE option is false
INFO:root:Flyway migration error log:
INFO:root - SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
INFO:root - SLF4J: Defaulting to no-operation (NOP) logger implementation
INFO:root - SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
INFO:root - Feb 10, 2023 10:55:33 PM net.snowflake.client.jdbc.internal.apache.arrow.memory.BaseAllocator <clinit>
INFO:root - INFO: Debug mode disabled.
INFO:root - Feb 10, 2023 10:55:33 PM net.snowflake.client.jdbc.internal.apache.arrow.memory.DefaultAllocationManagerOption getDefaultAllocationManagerFactory
INFO:root - INFO: allocation manager type not specified, using netty as the default type
INFO:root - Feb 10, 2023 10:55:33 PM net.snowflake.client.jdbc.internal.apache.arrow.memory.CheckAllocator reportResult
INFO:root - INFO: Using DefaultAllocationManager at memory/DefaultAllocationManagerFactory.class
INFO:root - Feb 10, 2023 10:55:33 PM net.snowflake.client.core.SFStatement executeQueryInternal
INFO:root - SEVERE: Exception creating result
INFO:root - java.lang.ExceptionInInitializerError
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.UnsafeAllocationManager.<clinit>(UnsafeAllocationManager.java:27)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.DefaultAllocationManagerFactory.<clinit>(DefaultAllocationManagerFactory.java:26)
INFO:root - 	at java.base/java.lang.Class.forName0(Native Method)
INFO:root - 	at java.base/java.lang.Class.forName(Unknown Source)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.DefaultAllocationManagerOption.getFactory(DefaultAllocationManagerOption.java:111)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.DefaultAllocationManagerOption.getDefaultAllocationManagerFactory(DefaultAllocationManagerOption.java:101)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.BaseAllocator$Config.getAllocationManagerFactory(BaseAllocator.java:735)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.ImmutableConfig.access$801(ImmutableConfig.java:24)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.ImmutableConfig$InitShim.getAllocationManagerFactory(ImmutableConfig.java:83)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.ImmutableConfig.<init>(ImmutableConfig.java:47)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.ImmutableConfig.<init>(ImmutableConfig.java:24)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.ImmutableConfig$Builder.build(ImmutableConfig.java:485)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.BaseAllocator.<clinit>(BaseAllocator.java:63)
INFO:root - 	at net.snowflake.client.jdbc.SnowflakeResultSetSerializableV1.create(SnowflakeResultSetSerializableV1.java:577)
INFO:root - 	at net.snowflake.client.jdbc.SnowflakeResultSetSerializableV1.create(SnowflakeResultSetSerializableV1.java:489)
INFO:root - 	at net.snowflake.client.core.SFResultSetFactory.getResultSet(SFResultSetFactory.java:29)
INFO:root - 	at net.snowflake.client.core.SFStatement.executeQueryInternal(SFStatement.java:218)
INFO:root - 	at net.snowflake.client.core.SFStatement.executeQuery(SFStatement.java:133)
INFO:root - 	at net.snowflake.client.core.SFStatement.execute(SFStatement.java:746)
INFO:root - 	at net.snowflake.client.core.SFStatement.execute(SFStatement.java:656)
INFO:root - 	at net.snowflake.client.jdbc.SnowflakeStatementV1.executeQueryInternal(SnowflakeStatementV1.java:236)
INFO:root - 	at net.snowflake.client.jdbc.SnowflakePreparedStatementV1.executeQuery(SnowflakePreparedStatementV1.java:118)
INFO:root - 	at org.flywaydb.core.internal.jdbc.JdbcTemplate.queryForString(JdbcTemplate.java:158)
INFO:root - 	at org.flywaydb.core.internal.database.snowflake.SnowflakeConnection.getCurrentSchemaNameOrSearchPath(SnowflakeConnection.java:51)
INFO:root - 	at org.flywaydb.core.internal.database.base.Connection.<init>(Connection.java:49)
INFO:root - 	at org.flywaydb.core.internal.database.snowflake.SnowflakeConnection.<init>(SnowflakeConnection.java:35)
INFO:root - 	at org.flywaydb.core.internal.database.snowflake.SnowflakeDatabase.doGetConnection(SnowflakeDatabase.java:63)
INFO:root - 	at org.flywaydb.core.internal.database.snowflake.SnowflakeDatabase.doGetConnection(SnowflakeDatabase.java:32)
INFO:root - 	at org.flywaydb.core.internal.database.base.Database.getConnection(Database.java:88)
INFO:root - 	at org.flywaydb.core.internal.database.base.Database.getMainConnection(Database.java:280)
INFO:root - 	at org.flywaydb.core.internal.schemahistory.SchemaHistoryFactory.prepareSchemas(SchemaHistoryFactory.java:58)
INFO:root - 	at org.flywaydb.core.FlywayExecutor.execute(FlywayExecutor.java:172)
INFO:root - 	at org.flywaydb.core.Flyway.migrate(Flyway.java:129)
INFO:root - 	at org.flywaydb.commandline.Main.executeOperation(Main.java:354)
INFO:root - 	at org.flywaydb.commandline.Main.main(Main.java:144)
INFO:root - Caused by: java.lang.RuntimeException: Failed to initialize MemoryUtil.
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.util.MemoryUtil.<clinit>(MemoryUtil.java:139)
INFO:root - 	... 35 more
INFO:root - Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio" to unnamed module @10adce03
INFO:root - 	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(Unknown Source)
INFO:root - 	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(Unknown Source)
INFO:root - 	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Unknown Source)
INFO:root - 	at java.base/java.lang.reflect.Field.setAccessible(Unknown Source)
INFO:root - 	at net.snowflake.client.jdbc.internal.apache.arrow.memory.util.MemoryUtil.<clinit>(MemoryUtil.java:87)
INFO:root - 	... 35 more
INFO:root - 
INFO:root - ERROR: Unable to determine the original schema for the connection
INFO:root - ----------------------------------------------------------
INFO:root - SQL State  : XX000
INFO:root - Error Code : 200001
INFO:root - Message    : JDBC driver internal error: exception creating result java.lang.ExceptionInInitializerError at net.snowflake.client.jdbc.internal.apache.arrow.memory.UnsafeAllocationManager.<clinit>(UnsafeAllocationManager.java:27).
INFO:root - 
INFO:root - Caused by: net.snowflake.client.jdbc.SnowflakeSQLException: JDBC driver internal error: exception creating result java.lang.ExceptionInInitializerError at net.snowflake.client.jdbc.internal.apache.arrow.memory.UnsafeAllocationManager.<clinit>(UnsafeAllocationManager.java:27).

**Comment by timo-aho** [comment]
Same happens with MacBook Pro macOS 13.2 Darwin 22.3.0 with Apple M1 Pro chip

**Comment by frichtarik** [comment]
@timo-aho thanks for the tip ... it did circumvent primary issue, however i'm now getting `Unable to parse statement in <versioned script>`  with `Caused by: Incomplete statement at line 3 col 1` 

this versioned file was OK for several years until now ... and is applied to 1000 + DBs, so i cannot edit it

**Comment by niegowic** [comment]
It a bug in new versions - https://github.com/flyway/flyway/pull/3614 - I created pull request - no one bother to review it.. I described problem here https://github.com/flyway/flyway/issues/3613 and seems root cause of the problem

**Comment by tjs145** [comment]
> @timo-aho thanks for the tip ... it did circumvent primary issue, however i'm now getting `Unable to parse statement in <versioned script>` with `Caused by: Incomplete statement at line 3 col 1`
> 
> this versioned file was OK for several years until now ... and is applied to 1000 + DBs, so i cannot edit it

We've now added the changes to the Snowflake parser. Does the error still happen in 9.15.1?

**Comment by frichtarik** [comment]
thankyou  but it did not fix everything 

now i have similar issue several schemas later in the deployment

ERROR: Unable to parse statement in /<path uninportant>/versioned/V1_0001__initial_schema.sql at line 230 col 1. See https://rd.gt/3ipi7Pm for more information: Incomplete statement at line 230 col 1: 
INFO:root: - 
INFO:root:- CREATE FILE FORMAT IF NOT EXISTS public.AVRO
INFO:root:- 	TYPE = AVRO
INFO:root:- COMMENT='This is the file format for importing Avro structured data.'
INFO:root:- ;
INFO:root:- Caused by: Incomplete statement at line 230 col 1: 
INFO:root:- 
INFO:root:- CREATE FILE FORMAT IF NOT EXISTS public.AVRO
INFO:root:- 	TYPE = AVRO
INFO:root:- COMMENT='This is the file format for importing Avro structured data.'
INFO:root:- ;

**Comment by frichtarik** [comment]
@tjs145 ^^ gentle reminder

**Comment by tjs145** [comment]
Does it work on 9.15.2? More changes were made to `CONDITIONALLY_CREATABLE_OBJECTS` in that version

**Comment by frichtarik** [comment]
again ... it fixed the issue from my last comment but there is still something further along the migration ... let me pull the log

**Comment by frichtarik** [comment]
with 9.15.2 i got stuck on aftermigrate.sql with following codeblock (internal identificators obfuscated)

```
INFO - begin transaction;
INFO - 
INFO - call schema.procedure_name(parse_json ('
INFO -     [
INFO -         {
INFO -             product_name: "product",
INFO -             logging_procedures_list: [
INFO -                 { log_label: "write_etl_log" }
INFO -             ]
INFO -         }
INFO -     ]'),
INFO -     current_database(), 
INFO -     $logging_db
INFO - );
INFO - 
INFO - commit;
```

**Comment by frichtarik** [comment]
updated name of the issue and versions in description of the issue

**Comment by iamtodor** [comment]
> On MacBook Pro, this was solved by #3559 That is, adding _&jdbc_query_result_format=json_ to the URL

For me this configuration does not work `flyway.url=jdbc:bigquery://https://www.googleapis.com/bigquery/v2:443&jdbc_query_result_format=json;ProjectId=<project-id>;OAuthType=0;OAuthServiceAcctEmail=service_account_name;OAuthPvtKeyPath=key/path/file.json;`

What am I missing in my command?

**Comment by frichtarik** [comment]
> > On MacBook Pro, this was solved by #3559 That is, adding _&jdbc_query_result_format=json_ to the URL
> 
> For me this configuration does not work `flyway.url=jdbc:bigquery://https://www.googleapis.com/bigquery/v2:443&jdbc_query_result_format=json;ProjectId=<project-id>;OAuthType=0;OAuthServiceAcctEmail=service_account_name;OAuthPvtKeyPath=key/path/file.json;`
> 
> What am I missing in my command?
> 
> Flyway was installed by [brew package](https://formulae.brew.sh/formula/flyway), the version is `Flyway Community Edition 9.15.2 by Redgate`

i believe solution is focused on Snowflake DB? no idea if applicable on other DB types

**Comment by frichtarik** [comment]
@tjs145 first, thanks for the great work 
but unfortunately i have to report another issue down the line

migration failed few schemas later with the same error `Incomplete statement at line` but i guess the reason is different 
flyway (9.16.1) does not seem to process escaped `'` correctly (in the `isn\'t`) considering it the end of statement argument

```
INFO:root-  comment on column SCHEMA.TABLE.COLUMN_NAME is
INFO:root- 	'{text:"Date and time a room closes and isn\'t available to join",
INFO:root- 	visible_ind:true,
INFO:root- 	deleted_ind:false}';
```


plus i've noticed following error in the sibling schema migration  (sibling - running in parallel with the first mentioned in this comment)

`ERROR: java.lang.NullPointerException: Cannot invoke "java.sql.ResultSet.getMetaData()" because "resultSet" is null`

unfortunately, there is nothing more in the log (this was last message for the schema) 
not sure if it's related

**Comment by frichtarik** [comment]
@tjs145 hi again. 
just rechecked with 9.16.3 and issues mentioned in previous comment are still reproducible

**Comment by tjs145** [comment]
> ```
> comment on column SCHEMA.TABLE.COLUMN_NAME is
> '{text:"Date and time a room closes and isn\'t available to join",
> visible_ind:true,
> deleted_ind:false}';
> ```

I don't get any error when running this statement on its own. 

The error might be caused by a previous statement in the same file. What's the minimum SQL required to reproduce the error?

**Comment by frichtarik** [comment]
@tjs145 
thank you for the patience, after long testing, i got to this minimal file content still throwing the issue 
it's caused by these 2 statements, i removed a lot of statements between them, there was no empty line at the end of file, I replaced only actual schema name with "SCHEMA"
```
comment on column SCHEMA.ROOM.END_TIME is
	'{text:"Date and time a room closes and isn\'t available to join",
	source_objects:["SCHEMA.END_TIME"]}';
comment on column SCHEMA.ROOM.ID is
	'{text:"Surrogate primary key for the table",
	source_objects:null}';
```
there is still no obvious cause of the issue for me

**Comment by tjs145** [comment]
Thanks @frichtarik - There was indeed a problem with the escaped `'`. 

There will be a fix for this in the next release and hopefully that unblocks you.

**Comment by frichtarik** [comment]
fix confirmed, thank you again
